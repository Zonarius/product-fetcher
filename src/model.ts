import { Observable } from "rxjs";

export interface Fetcher<T extends Record<string, any>> {
  matches(url: string): boolean;
  fetchProducts(url: string): Observable<T>
}
