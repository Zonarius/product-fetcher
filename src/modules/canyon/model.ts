export interface CanyonProduct {
  name: string;
  price: number;
  weight: number;
  url: string;
}