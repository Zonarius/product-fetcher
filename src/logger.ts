import log4js from 'log4js';

log4js.configure({
  appenders: {
    out: { type: 'stderr' }
  },
  categories: {
    default: { appenders: ["out"], level: "info" }
  }
});
export const logger = log4js.getLogger();
