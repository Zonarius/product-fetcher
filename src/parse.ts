export function parseNumber(input: string | undefined): number | undefined {
  if (!input) {
    return undefined;
  }
  const match = input.match(/[\d,\.]+/);
  if (match) {
    const found = match[0];
    let result = "";
    let commaFound = false;
    for (let i = found.length - 1; i >= 0; i--) {
      if (found[i] === "," || found[i] === ".") {
        if (!commaFound) {
          result = "." + result;
          commaFound = true;
        }
      } else {
        result = found[i] + result;
      }
    }
    return Number(result);
  } else {
    return undefined;
  }
}