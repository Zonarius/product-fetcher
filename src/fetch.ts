import fetch from 'node-fetch';
import { HTMLElement, parse } from 'node-html-parser';
import { defer, from, map, Observable } from 'rxjs';
import { logger } from './logger';

export function getText(url: string): Promise<string> {
  logger.info(`Getting ${url}`);
  return fetch(url).then(r => r.text());
}

export function getParsed(url: string): Promise<HTMLElement> {
  return getText(url).then(parse);
}

export function getParsedObs(url: string): Observable<HTMLElement> {
  return defer(() => from(getParsed(url)));
}