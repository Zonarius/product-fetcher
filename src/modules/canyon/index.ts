import { HTMLElement } from "node-html-parser";
import { from, mergeMap, Observable, of, switchMap } from "rxjs";
import { getParsedObs } from "../../fetch";
import { Fetcher } from "../../model";
import { parseNumber } from "../../parse";
import { createRecord } from "../../util";
import { CanyonProduct } from "./model";

export function matches(url: string) {
  return /^https?:\/\/www\.canyon\.com/.test(url);
}

export function fetchProducts(url: string): Observable<CanyonProduct> {
  return getParsedObs(url).pipe(
    switchMap(document => {
      if (document.querySelector("#section-product-grid")) {
        return from(getProductListUrls(document))
          .pipe(
            mergeMap(fetchProducts)
          )
      } else {
        return of({ ...parseProduct(document), url } as CanyonProduct);
      }
    })
  )
}

export function getProductListUrls(document: HTMLElement): string[] {
  return document.querySelectorAll("a.productTile__link")
    .map(el => el.getAttribute("href")!);
}

export function parseProduct(document: HTMLElement): Partial<CanyonProduct> {
  return {
    name: document.querySelector("h1.productDescription__productName").rawText.trim(),
    price: parseNumber(document.querySelector(".productDescription__priceSale").rawText)!,
    weight: getWeight(document),
  }
}

function getWeight(document: HTMLElement): number {
  return parseNumber(getCharacteristics(document)["Weight"])!;
}

function getCharacteristics(document: HTMLElement): Record<string, string> {
  return createRecord(
    document.querySelectorAll(".productDescription__characteristicsListItem")
      .map(item => {
        const [label, detail] = item.querySelectorAll("div");
        return [label.rawText.trim(), detail.rawText.trim()];
      })
  );
}

const moduletypeCheck = { matches, fetchProducts } as Fetcher<CanyonProduct>