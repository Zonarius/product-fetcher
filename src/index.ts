#!/usr/bin/env -S node -r "ts-node/register"

import { Command, Option } from 'commander';
import { take, toArray } from 'rxjs';
import { logger } from './logger';
import modules from './modules';

const { version } = require("../package.json");
const program = new Command();

process.env["NODE_OPTIONS"] = "--insecure-http-parser";

program
  .name("product-fetcher")
  .description("Fetches product data from various sources.")
  .version(version)
  .option("-v, --verbose", "Verbose output")
  .option("--list-sources", "Lists all available sources")
  .option("-l, --limit <limit>", "The maximum number of products to fetch", Number, -1)
  .argument("[url]", "The url to fetch products from")
  .action(main)
  .parse();

async function main(url: string) {
  const options = program.opts();
  logger.level = options.verbose
    ? "debug"
    : "info";
    
  if (options.listSources) {
    Object.keys(modules).forEach(m => console.log(m));
    return;
  }

  if (!url) {
    program.help({ error: true })
  }

  const matched = Object.values(modules).find(f => f.matches(url));
  if (!matched) {
    logger.error("No module matches the URL!");
    return;
  }

  const obs = options.limit >= 0
    ? matched.fetchProducts(url).pipe(take(options.limit))
    : matched.fetchProducts(url);

  obs.pipe( toArray() )
    .subscribe(products => {
      console.log(JSON.stringify({ products }));
    })
}