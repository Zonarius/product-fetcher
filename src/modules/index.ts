import { Fetcher } from '../model';
import * as decathlon from './decathlon';
import * as canyon from './canyon';

export default {
  decathlon, canyon
} as Record<string, Fetcher<any>>