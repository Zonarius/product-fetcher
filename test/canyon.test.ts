import parse from "node-html-parser";
import { getText } from "../src/fetch"
import { getProductListUrls, parseProduct } from "../src/modules/canyon";
import { testCache } from "./cache";

describe("canyon", () => {
  test("parse single product", async () => {
    const doc = await getParsed("https://www.canyon.com/en-at/road-bikes/endurance-bikes/endurace/cf-sl/endurace-cf-sl-8-disc-di2/2949.html");
    const parsed = parseProduct(doc);
    expect(parsed).toMatchSnapshot();
  })

  test("parse product list", async () => {
    const doc = await getParsed("https://www.canyon.com/en-at/road-bikes/endurance-bikes/endurace/cf-sl/");
    expect(getProductListUrls(doc).length).toBeGreaterThan(0);
  })
})

async function getParsed(url: string) {
  const text = await testCache(() => getText(url));
  return parse(text);
}