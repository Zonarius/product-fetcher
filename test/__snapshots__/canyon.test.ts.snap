// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`canyon parse single product 1`] = `
Object {
  "name": "Endurace CF SL 8 Disc Di2",
  "price": 3599,
  "weight": 7.94,
}
`;
